Schema update process only runs on Windows

### Description
Running the schema update process with `mvn clean install -P process-schemas` on a Unix-like platform results in the following exception:

```
Caused by: java.lang.IllegalArgumentException: /home/magnusbk/src/eiffel-remrem-semantics/src\main\resources\schemas\input does not exist
    at org.apache.commons.io.FileUtils.verifiedListFiles (FileUtils.java:1741)
    at org.apache.commons.io.FileUtils.cleanDirectory (FileUtils.java:1716)
    at com.ericsson.eiffel.remrem.semantics.schemas.LocalRepo.readSchemas (LocalRepo.java:46)
    at com.ericsson.eiffel.remrem.semantics.clone.PrepareLocalEiffelSchemas.main (PrepareLocalEiffelSchemas.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:294)
    at java.lang.Thread.run (Thread.java:748)
```

This is unsurprising given the backslash-separated paths in [EiffelConstants](https://github.com/eiffel-community/eiffel-remrem-semantics/blob/07d8ab4f7bcaf6810edc316b990a0337cc81806e/src/main/java/com/ericsson/eiffel/remrem/semantics/schemas/EiffelConstants.java).

### Motivation
The schema update process should obviously be possible to run on platforms other than Windows.

### Exemplification
I'd like to be able to update the schemas from a Linux host.

### Benefits
Obvious.

### Possible Drawbacks
None.

--- a/pom.xml
+++ b/pom.xml
@@ -43,6 +43,11 @@
             <scope>compile</scope>
         </dependency>
         <dependency>
+            <groupId>com.vdurmont</groupId>
+            <artifactId>semver4j</artifactId>
+            <version>3.1.0</version>
+        </dependency>
+        <dependency>
             <groupId>org.glassfish.hk2.external</groupId>
             <artifactId>javax.inject</artifactId>
         </dependency>

--- a/src/main/java/com/ericsson/eiffel/remrem/semantics/clone/PrepareLocalEiffelSchemas.java
+++ b/src/main/java/com/ericsson/eiffel/remrem/semantics/clone/PrepareLocalEiffelSchemas.java
@@ -23,6 +23,8 @@ import java.net.Proxy;
 import java.net.ProxySelector;
 import java.net.SocketAddress;
 import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
@@ -64,17 +66,17 @@ public class PrepareLocalEiffelSchemas {
      * @param localEiffelRepoPath
      *            destination path to clone the repo.
      */
-    private void cloneEiffelRepo(final String repoURL, final String branch, final File localEiffelRepoPath) {
+    private void cloneEiffelRepo(final String repoURL, final String branch, final Path localEiffelRepoPath) {
         Git localGitRepo = null;
         
         // checking for repository exists or not in the localEiffelRepoPath
         try {
-            if (!localEiffelRepoPath.exists()) {
+            if (Files.notExists(localEiffelRepoPath)) {
                 // cloning github repository by using URL and branch name into local
-                localGitRepo = Git.cloneRepository().setURI(repoURL).setBranch("master").setDirectory(localEiffelRepoPath).call();
+                localGitRepo = Git.cloneRepository().setURI(repoURL).setBranch("master").setDirectory(localEiffelRepoPath.toFile()).call();
             } else {
                 // If required repository already exists
-                localGitRepo = Git.open(localEiffelRepoPath);
+                localGitRepo = Git.open(localEiffelRepoPath.toFile());
 
                 // Reset to normal if uncommitted changes are present
                 localGitRepo.reset().call();
@@ -152,12 +154,12 @@ public class PrepareLocalEiffelSchemas {
      * @param eiffelRepoPath
      *            local eiffel repository url
      */
-    private void copyOperationSchemas(final String operationsRepoPath, final String eiffelRepoPath) {
-        final File operationSchemas = new File(operationsRepoPath + File.separator + EiffelConstants.SCHEMA_LOCATION);
-        final File eiffelSchemas = new File(eiffelRepoPath + File.separator + EiffelConstants.SCHEMA_LOCATION);
-        if (operationSchemas.isDirectory()) {
+    private void copyOperationSchemas(final Path operationsRepoPath, final Path eiffelRepoPath) {
+        final Path operationSchemas = operationsRepoPath.resolve(EiffelConstants.SCHEMA_LOCATION);
+        final Path eiffelSchemas = eiffelRepoPath.resolve(EiffelConstants.SCHEMA_LOCATION);
+        if (Files.isDirectory(operationSchemas)) {
             try {
-                FileUtils.copyDirectory(operationSchemas, eiffelSchemas);
+                FileUtils.copyDirectory(operationSchemas.toFile(), eiffelSchemas.toFile());
             } catch (IOException e) {
                 System.out.println("Exception occurred while copying schemas from operations repository to eiffel repository");
                 e.printStackTrace();
@@ -176,9 +178,8 @@ public class PrepareLocalEiffelSchemas {
         final String operationRepoUrl = args[2];
         final String operationRepoBranch = args[3];
 
-        final File localEiffelRepoPath = new File(System.getProperty(EiffelConstants.USER_HOME) + File.separator + EiffelConstants.EIFFEL);
-        final File localOperationsRepoPath = new File(
-                System.getProperty(EiffelConstants.USER_HOME) + File.separator + EiffelConstants.OPERATIONS_REPO_NAME);
+        final Path localEiffelRepoPath = EiffelConstants.USER_HOME.resolve(EiffelConstants.EIFFEL);
+        final Path localOperationsRepoPath = EiffelConstants.USER_HOME.resolve(EiffelConstants.OPERATIONS_REPO_NAME);
 
         // Clone Eiffel Repo from GitHub 
         prepareLocalSchema.cloneEiffelRepo(eiffelRepoUrl, eiffelRepoBranch, localEiffelRepoPath);
@@ -187,7 +188,7 @@ public class PrepareLocalEiffelSchemas {
         prepareLocalSchema.cloneEiffelRepo(operationRepoUrl, operationRepoBranch, localOperationsRepoPath);
 
         //Copy operations repo Schemas to location where Eiffel repo schemas available
-        prepareLocalSchema.copyOperationSchemas(localOperationsRepoPath.getAbsolutePath(), localEiffelRepoPath.getAbsolutePath());
+        prepareLocalSchema.copyOperationSchemas(localOperationsRepoPath.toAbsolutePath(), localEiffelRepoPath.toAbsolutePath());
 
         // Read and Load JsonSchemas from Cloned Directory 
         final LocalRepo localRepo = new LocalRepo(localEiffelRepoPath);

--- a/src/main/java/com/ericsson/eiffel/remrem/semantics/clone/PrepareLocalEiffelSchemas.java
+++ b/src/main/java/com/ericsson/eiffel/remrem/semantics/clone/PrepareLocalEiffelSchemas.java
@@ -14,7 +14,6 @@
 */
 package com.ericsson.eiffel.remrem.semantics.clone;
 
-import java.io.File;
 import java.io.IOException;
 import java.net.Authenticator;
 import java.net.InetSocketAddress;
@@ -25,10 +24,10 @@ import java.net.SocketAddress;
 import java.net.URI;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Locale;
+import java.util.Map;
 import java.util.ResourceBundle;
 
 import org.apache.commons.io.FileUtils;
@@ -194,18 +193,12 @@ public class PrepareLocalEiffelSchemas {
         final LocalRepo localRepo = new LocalRepo(localEiffelRepoPath);
         localRepo.readSchemas();
 
-        final ArrayList<String> jsonEventNames = localRepo.getJsonEventNames();
-        final ArrayList<File> jsonEventSchemas = localRepo.getJsonEventSchemas();
-
-        // Schema changes 
+        // Schema changes
         final SchemaFile schemaFile = new SchemaFile();
 
-        // Iterate the Each jsonSchema file to Add and Modify the necessary properties 
-        if (jsonEventNames != null && jsonEventSchemas != null) {
-            for (int i = 0; i < jsonEventNames.size(); i++) {
-                schemaFile.modify(jsonEventSchemas.get(i), jsonEventNames.get(i));
-            }
+        // Iterate over available input schemas and create new and patched files
+        for (Map.Entry<String, Path> event : localRepo.getJsonEventSchemas().entrySet()) {
+             schemaFile.modify(event.getValue().toFile(), event.getKey());
         }
-
     }
 }

--- a/src/main/java/com/ericsson/eiffel/remrem/semantics/schemas/EiffelConstants.java
+++ b/src/main/java/com/ericsson/eiffel/remrem/semantics/schemas/EiffelConstants.java
@@ -14,6 +14,9 @@
 */
 package com.ericsson.eiffel.remrem.semantics.schemas;
 
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
 /**
  * This is a constants class used in other classes
  * 
@@ -41,13 +44,13 @@ public final class EiffelConstants {
     public static String COM_ERICSSON_EIFFEL_SEMANTICS_EVENTS = "com.ericsson.eiffel.semantics.events.";
     public static String TYPE = "type";
     public static String OBJECTTYPE = "object";
-    public static String INPUT_EIFFEL_SCHEMAS = "src\\main\\resources\\schemas\\input";
+    public static Path INPUT_EIFFEL_SCHEMAS = Paths.get("src", "main", "resources", "schemas", "input");
     public static String EIFFEL = "eiffel";
     public static String OPERATIONS_REPO_NAME = "eiffel-operations-extension";
-    public static String USER_DIR = System.getProperty("user.dir");
-    public static String SCHEMA_LOCATION = "\\schemas";
+    public static Path USER_DIR = Paths.get(System.getProperty("user.dir"));
+    public static Path SCHEMA_LOCATION = Paths.get("schemas");
     public static String JSON_MIME_TYPE = ".json";
-    public static String USER_HOME = "user.home";
+    public static Path USER_HOME = Paths.get(System.getProperty("user.home"));
     public static String ACTIVITY = "activity";
     public static String ARTIFACT = "artifact";
     public static String SERVICE = "service";

--- a/src/main/java/com/ericsson/eiffel/remrem/semantics/schemas/LocalRepo.java
+++ b/src/main/java/com/ericsson/eiffel/remrem/semantics/schemas/LocalRepo.java
@@ -16,6 +16,10 @@ package com.ericsson.eiffel.remrem.semantics.schemas;
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.ArrayList;
 
 import org.apache.commons.io.FileUtils;
@@ -29,10 +33,10 @@ import org.apache.commons.io.FileUtils;
  */
 public class LocalRepo {
 	private ArrayList<File> jsonEventSchemas;
-	private File localSchemasPath;
+	private Path localSchemasPath;
 	private ArrayList<String> jsonEventNames;
 
-	public LocalRepo(File localSchemasPath) {
+	public LocalRepo(Path localSchemasPath) {
 		this.localSchemasPath = localSchemasPath;
 	}
 
@@ -41,16 +45,15 @@ public class LocalRepo {
 	 * Repo
 	 */
 
-	public void readSchemas() {
+	public void readSchemas() throws IOException {
 		try {
-			FileUtils.cleanDirectory(new File(EiffelConstants.USER_DIR + File.separator + EiffelConstants.INPUT_EIFFEL_SCHEMAS));
+			FileUtils.cleanDirectory(EiffelConstants.USER_DIR.resolve(EiffelConstants.INPUT_EIFFEL_SCHEMAS).toFile());
 		} catch (IOException e) {
 			e.printStackTrace();
 		}
 		jsonEventNames = new ArrayList<String>();
 		jsonEventSchemas = new ArrayList<File>();
-		String filePath = localSchemasPath + EiffelConstants.SCHEMA_LOCATION;
-		loadEiffelSchemas(filePath, "");
+		loadEiffelSchemas(localSchemasPath.resolve(EiffelConstants.SCHEMA_LOCATION), Paths.get(""));
 	}
 
 	/**
@@ -65,15 +68,15 @@ public class LocalRepo {
 	 *            event name.
 	 * 
 	 */
-	private void loadEiffelSchemas(String jsonFilePath, String directoryName) {
-		File file = new File(jsonFilePath);
-		File[] files = file.listFiles();
-		for (File jsonFile : files) {
-			if (jsonFile.isDirectory()) {
-				loadEiffelSchemas(jsonFile.getAbsolutePath(), jsonFile.getName());
-			} else {
-				jsonEventNames.add(directoryName);
-				jsonEventSchemas.add(jsonFile);
+	private void loadEiffelSchemas(Path jsonFilePath, Path directoryName) throws IOException {
+		try (DirectoryStream<Path> dirStream = Files.newDirectoryStream(jsonFilePath)) {
+			for (Path jsonFile : dirStream) {
+				if (Files.isDirectory(jsonFile)) {
+					loadEiffelSchemas(jsonFile.toAbsolutePath(), jsonFile.getFileName());
+				} else {
+					jsonEventNames.add(directoryName.toString());
+					jsonEventSchemas.add(jsonFile.toFile());
+				}
 			}
 		}
 	}

--- a/src/main/java/com/ericsson/eiffel/remrem/semantics/schemas/LocalRepo.java
+++ b/src/main/java/com/ericsson/eiffel/remrem/semantics/schemas/LocalRepo.java
@@ -14,15 +14,16 @@
 */
 package com.ericsson.eiffel.remrem.semantics.schemas;
 
-import java.io.File;
+import com.vdurmont.semver4j.Semver;
 import java.io.IOException;
 import java.nio.file.DirectoryStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.nio.file.Paths;
-import java.util.ArrayList;
-
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.StreamSupport;
 import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.FilenameUtils;
 
 /**
  * This class is used to Iterate over the Eiffel schemas that are cloned from
@@ -32,9 +33,8 @@ import org.apache.commons.io.FileUtils;
  *
  */
 public class LocalRepo {
-	private ArrayList<File> jsonEventSchemas;
+	private final Map<String, Path> jsonEventSchemas = new HashMap<>();
 	private Path localSchemasPath;
-	private ArrayList<String> jsonEventNames;
 
 	public LocalRepo(Path localSchemasPath) {
 		this.localSchemasPath = localSchemasPath;
@@ -51,9 +51,7 @@ public class LocalRepo {
 		} catch (IOException e) {
 			e.printStackTrace();
 		}
-		jsonEventNames = new ArrayList<String>();
-		jsonEventSchemas = new ArrayList<File>();
-		loadEiffelSchemas(localSchemasPath.resolve(EiffelConstants.SCHEMA_LOCATION), Paths.get(""));
+		loadEiffelSchemas(localSchemasPath.resolve(EiffelConstants.SCHEMA_LOCATION));
 	}
 
 	/**
@@ -63,32 +61,26 @@ public class LocalRepo {
 	 * @param jsonFilePath
 	 *            - This parameter is used to pass Location of the Schemas
 	 *            Directory
-	 * @param directoryName
-	 *            - This parameter is used to rename the File with corresponding
-	 *            event name.
-	 * 
 	 */
-	private void loadEiffelSchemas(Path jsonFilePath, Path directoryName) throws IOException {
-		try (DirectoryStream<Path> dirStream = Files.newDirectoryStream(jsonFilePath)) {
-			for (Path jsonFile : dirStream) {
-				if (Files.isDirectory(jsonFile)) {
-					loadEiffelSchemas(jsonFile.toAbsolutePath(), jsonFile.getFileName());
-				} else {
-					jsonEventNames.add(directoryName.toString());
-					jsonEventSchemas.add(jsonFile.toFile());
+	private void loadEiffelSchemas(final Path jsonFilePath) throws IOException {
+		try (DirectoryStream<Path> schemaDirStream =
+					 Files.newDirectoryStream(jsonFilePath, Files::isDirectory)) {
+			for (Path eventDir : schemaDirStream) {
+				try (DirectoryStream<Path> eventDirStream =
+							 Files.newDirectoryStream(eventDir, file -> file.toString().endsWith(".json"))) {
+					// Turn the filenames into versions and find the greatest version
+					Semver latestSchemaVersion = StreamSupport.stream(eventDirStream.spliterator(), false)
+							.map(path -> new Semver(FilenameUtils.removeExtension(path.getFileName().toString())))
+							.max(Semver::compareTo)
+							.get();
+					jsonEventSchemas.put(eventDir.getFileName().toString(),
+							eventDir.resolve(latestSchemaVersion.toString() + ".json"));
 				}
 			}
 		}
 	}
 
-	public ArrayList<File> getJsonEventSchemas() {
+	public Map<String, Path> getJsonEventSchemas() {
 		return jsonEventSchemas;
 	}
-
-	public ArrayList<String> getJsonEventNames() {
-		return jsonEventNames;
-	}
-	
-	
-
 }

--- a/src/main/java/com/ericsson/eiffel/remrem/semantics/schemas/SchemaFile.java
+++ b/src/main/java/com/ericsson/eiffel/remrem/semantics/schemas/SchemaFile.java
@@ -17,6 +17,7 @@ package com.ericsson.eiffel.remrem.semantics.schemas;
 import java.io.File;
 import java.io.FileWriter;
 import java.nio.file.Files;
+import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Iterator;
 import java.util.Map.Entry;
@@ -236,16 +237,15 @@ public class SchemaFile {
      *            an input parameter to this method
      */
     public void createNewInputJsonSchema(String jsonFileName, JsonObject jsonObject) {
-        String currentWorkingDir = EiffelConstants.USER_DIR;
         FileWriter writer = null;
-        String copyFilePath = currentWorkingDir + File.separator + EiffelConstants.INPUT_EIFFEL_SCHEMAS;
-        String newFileName = copyFilePath + File.separator + jsonFileName + EiffelConstants.JSON_MIME_TYPE;
+        Path newFileName = EiffelConstants.USER_DIR.resolve(EiffelConstants.INPUT_EIFFEL_SCHEMAS)
+                .resolve(jsonFileName + EiffelConstants.JSON_MIME_TYPE);
         Gson gson = new GsonBuilder().setPrettyPrinting().create();
         JsonParser jp = new JsonParser();
         JsonElement je = jp.parse(jsonObject.toString());
         String prettyJsonString = gson.toJson(je);
         try {
-            writer = new FileWriter(newFileName);
+            writer = new FileWriter(newFileName.toFile());
             writer.write(prettyJsonString);
         } catch (Exception e) {
             e.printStackTrace();

